python/ql/src/experimental/Security/CWE-022bis/TarSlipImprov.ql

```diff
@@ -0,0 +1,116 @@
+ */
+class AllTarfileOpens extends API::CallNode {
+  AllTarfileOpens() {
+    this = tarfileOpen().getACall()
+    or
+    exists(API::Node closing, Node arg |
+      closing = API::moduleImport("contextlib").getMember("closing") and
+      this = closing.getACall() and
+      arg = this.getArg(0) and
+      arg = tarfileOpen().getACall()
+    )
+  }
+}
+/**
+ * A taint-tracking configuration for detecting more "TarSlip" vulnerabilities.
+ */
+class Configuration extends TaintTracking::Configuration {
+  Configuration() { this = "TarSlip" }
+
+  override predicate isSource(DataFlow::Node source) { source = tarfileOpen().getACall() }
+
+  override predicate isSink(DataFlow::Node sink) {
+    // A sink capturing method calls to `extractall` without `members` argument.
+    // For a call to `file.extractall` without `members` argument, `file` is considered a sink.
+    exists(MethodCallNode call , AllTarfileOpens atfo|
+      call = atfo.getReturn().getMember("extractall").getACall() and
+      not exists(Node arg | arg = call.getArgByName("members")) and
+      sink = call.getObject()
+    )
+    or
+    // A sink capturing method calls to `extractall` with `members` argument.
+    // For a call to `file.extractall` with `members` argument, `file` is considered a sink if not
+    // a the `members` argument contains a NameConstant as None, a List or call to the method `getmembers`.
+    // Otherwise, the argument of `members` is considered a sink.
+    exists(MethodCallNode call, Node arg, AllTarfileOpens atfo|
+      call = atfo.getReturn().getMember("extractall").getACall() and
+      arg = call.getArgByName("members") and
+      if
+        arg.asCfgNode() instanceof NameConstantNode or
+        arg.asCfgNode() instanceof ListNode
+      then sink = call.getObject()
+      else
+        if arg.(MethodCallNode).getMethodName() = "getmembers"
+        then sink = arg.(MethodCallNode).getObject()
+        else sink = call.getArgByName("members")
+    )
+    or
+    // An argument to `extract` is considered a sink.
+    exists(AllTarfileOpens atfo | sink = atfo.getReturn().getMember("extract").getACall().getArg(0))
+    or
+    //An argument to `_extract_member` is considered a sink.
+    exists(MethodCallNode call, AllTarfileOpens atfo |
+      call = atfo.getReturn().getMember("_extract_member").getACall() and
+      call.getArg(1).(AttrRead).accesses(sink, "name")
+    )
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+    exists(AttrRead attr, MethodCallNode call |
+      attr.accesses(nodeFrom, "getmembers") and
+      nodeFrom = call.getObject() and
+      nodeFrom instanceof AllTarfileOpens and
+      nodeTo = call
+    )
+    or
+    exists(DataFlow::CallCfgNode closing |
+      closing = API::moduleImport("contextlib").getMember("closing").getACall() and
+      nodeFrom = closing.getArg(0) and
+      nodeFrom = tarfileOpen().getACall() and
+      nodeTo = closing
+    )
```

Comment:
Good, this is the step I wanted. However, now it is only added if the argument looks like an `open`-call, and we would like to also handle cases like
```python
f = tar file.open(...)
...
with closing(f) as t:
   ...
```
The data flow will be able to find this path, except the crucial step from `f` to `t` will not be there.
To fix this, while still restricting the argument, we can use the `API::CallNode` (which extends `DataFlowCallCfgNode`):
```suggestion
    exists(API::CallNode closing |
      closing = API::moduleImport("contextlib").getMember("closing").getACall() and
      nodeFrom = closing.getArg(0) and
      nodeFrom = tarfileOpen().getReturn().getAValueReachingSink() and
      nodeTo = closing
    )
```